src/awesome/button.rs

```diff
@@ -0,0 +1,343 @@
+use std::fmt::{self, Display, Formatter};
+use std::default::Default;
+use std::rc::Rc;
+use rlua::{self, Table, Lua, UserData, ToLua, Value};
+use super::object::{Object, Objectable};
+use super::signal;
+use super::property::Property;
+use super::class::{self, Class};
+use rustwlc::types::KeyMod;
+use xcb::ffi::xproto::xcb_button_t;
+
+#[derive(Clone, Debug)]
+pub struct ButtonState {
+    button: xcb_button_t,
+    modifiers: KeyMod
+}
+
+#[derive(Clone, Debug)]
+pub struct Button<'lua>(Table<'lua>);
+
+impl <'lua> Button<'lua> {
+    pub fn button(&self) -> rlua::Result<Value<'lua>> {
+        let button = self.0.get::<_, ButtonState>("data")?;
+        Ok(Value::Integer(button.button as _))
+    }
+
+    pub fn set_button(&self, new_val: xcb_button_t) -> rlua::Result<()> {
+        let mut button = self.0.get::<_, ButtonState>("data")?;
+        button.button = new_val;
+        self.0.set("data", button)?;
+        Ok(())
+    }
+
+    pub fn modifiers(&self) -> rlua::Result<KeyMod> {
+        let button = self.0.get::<_, ButtonState>("data")?;
+        Ok(button.modifiers)
+    }
+
+    pub fn set_modifiers(&self, mods: Table<'lua>) -> rlua::Result<()> {
+        use ::lua::mods_to_rust;
+        let mut button = self.0.get::<_, ButtonState>("data")?;
+        button.modifiers = mods_to_rust(mods)?;
+        self.0.set("data", button)?;
+        Ok(())
+    }
+}
+
+impl <'lua> ToLua<'lua> for Button<'lua> {
+    fn to_lua(self, lua: &'lua Lua) -> rlua::Result<Value<'lua>> {
+        self.0.to_lua(lua)
+    }
+}
+
+impl <'lua> Objectable<'lua, Button<'lua>, ButtonState> for Button<'lua> {
+    fn _wrap(table: Table<'lua>) -> Button {
+        Button(table)
+    }
+
+    fn get_table(self) -> Table<'lua> {
+        self.0
+    }
+}
+
+impl Display for ButtonState {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "Button: {:p}", self)
+    }
+}
+
+impl Default for ButtonState {
+    fn default() -> Self {
+        ButtonState {
+            button: xcb_button_t::default(),
+            modifiers: KeyMod::empty()
+        }
+    }
+}
+
+impl UserData for ButtonState {}
+
+/// Makes a new button stored in a table beside its signals
+pub fn allocator(lua: &Lua) -> rlua::Result<Object> {
```

Comment:
Why not make this a static method on `Button`? E.g `Button::allocate`?

Reply:
Done, but it might be better on Objectable...I won't move it yet, but time will tell...